lib/Crypto/importkeyrequest.cpp

```diff
@@ -0,0 +1,227 @@
+ */
+void ImportKeyRequest::setInteractionParameters(
+        const Sailfish::Crypto::InteractionParameters &uiParams)
+{
+    Q_D(ImportKeyRequest);
+    if (d->m_status != Request::Active && d->m_uiParams != uiParams) {
+        d->m_uiParams = uiParams;
+        if (d->m_status == Request::Finished) {
+            d->m_status = Request::Inactive;
+            emit statusChanged();
+        }
+        emit interactionParametersChanged();
+    }
+}
+
+/*!
+ * \brief Returns the key which should be imported.
+ */
+Key ImportKeyRequest::key() const
+{
+    Q_D(const ImportKeyRequest);
+    return d->m_key;
+}
+
+/*!
+ * \brief Sets the \a key which should be imported.
+ */
+void ImportKeyRequest::setKey(const Key &key)
+{
+    Q_D(ImportKeyRequest);
+    if (d->m_status != Request::Active && d->m_key != key) {
+        d->m_key = key;
+        if (d->m_status == Request::Finished) {
+            d->m_status = Request::Inactive;
+            emit statusChanged();
+        }
+        emit keyChanged();
+    }
+}
+
+/*!
+ * \brief Returns the imported key
+ *
+ * Note: this value is only valid if the status of the request is Request::Finished.
+ */
+Key ImportKeyRequest::importedKey() const
+{
+    Q_D(const ImportKeyRequest);
+    return d->m_importedKey;
+}
+
+Request::Status ImportKeyRequest::status() const
+{
+    Q_D(const ImportKeyRequest);
+    return d->m_status;
+}
+
+Result ImportKeyRequest::result() const
+{
+    Q_D(const ImportKeyRequest);
+    return d->m_result;
+}
+
+CryptoManager *ImportKeyRequest::manager() const
+{
+    Q_D(const ImportKeyRequest);
+    return d->m_manager.data();
+}
+
+void ImportKeyRequest::setManager(CryptoManager *manager)
+{
+    Q_D(ImportKeyRequest);
+    if (d->m_manager.data() != manager) {
+        d->m_manager = manager;
+        emit managerChanged();
+    }
+}
+
+void ImportKeyRequest::startRequest()
+{
+    Q_D(ImportKeyRequest);
+    if (d->m_status != Request::Active && !d->m_manager.isNull()) {
+        d->m_status = Request::Active;
+        emit statusChanged();
+        if (d->m_result.code() != Result::Pending) {
+            d->m_result = Result(Result::Pending);
+            emit resultChanged();
+        }
+
+        QDBusPendingReply<Result, Key> reply =
+                d->m_manager->d_ptr->importKey(d->m_key,
+                                               d->m_uiParams,
+                                               d->m_cryptoPluginName);
+        if (reply.isError()) {
+            d->m_status = Request::Finished;
+            d->m_result = Result(Result::CryptoManagerNotInitialisedError,
+                                 reply.error().message());
+            emit statusChanged();
+            emit resultChanged();
+        } else if (reply.isFinished()
+                // work around a bug in QDBusAbstractInterface / QDBusConnection...
+                && reply.argumentAt<0>().code() != Sailfish::Crypto::Result::Succeeded) {
+            qDebug() << "Why?";
```

Comment:
Stray debug ... or philosophizing ;-)

Reply:
A bit of both. I'm certainly struggling philosophically with an asynchronous call that's returning a result synchronously but it is stray debugging code I failed to clean up before committing.